### ISSUE TITLE
Option to defer lint on the current line

### ISSUE THREAD

**Issue by posita** [atom-community/linter#749]
Reopened from #170:

@sandstorm wrote:
>It would be neat if the current row was deferred/ignored. I find that linting errors are frequently shown while coding, which is only distracting. I.e. I'm in the midst of writing a line of code and it shows an error because I have not finished typing yet.
>
>It teaches users to pay less attention to the highlight, which goes against the idea of linting. Perhaps it could keep the changes on the current line ignored until the cursor moves to another row. That should avoid some needless linting.
>
>Only increasing the 'lint on change' interval is not enough. Since a too high value will have lint errors 'linger', and it's better if they are immediately removed when a lint issue is resolved. Perhaps another solution would be to add delay to showing errors (~10 seconds), but less delay to removing them (~1 second).
>
>What are your thoughts on this?

@posita wrote:
>A more nuanced (but probably more difficult to implement) approach would be to provide an option to ignore a *particular* warning in the current context rather than all warnings (consider, e.g., `# pylint: disable=...[,...]` for `pylint`, etc.). A list of potential warnings could be built from those still presenting themselves on that line. I'm guessing this would have the be integrated with the various implementing packages (where possible).

**Comment by steelbrain** [comment]
We've had this conversation before and it was decided that disabling messages for lines is something that the linter providers should handle instead of the core linter, also the latest linter uses core's onDidStopChanging event and it actually invalidates the markers when that text is removed, and You can also choose to disable lintOnFly.

**Comment by adambuczynski** [comment]
@steelbrain Could you tell us a bit more about why you decided the core linter should not be in charge of it? I think it would be nicer if the core linter would handle this, as presumably it's the last stop before the actual messages get shown, so it'd be the perfect place to centralize this kind of behavior. 

I've looked at the two linter packages that I currently use (jshint and jscs), and neither of them have the option to defer linting for the current line or until you've stopped typing. 

So instead of having to go to every individual package maintainer and ask for the same feature everywhere, wouldn't it make more sense to have it centralized?

Happy to discuss further and hear your thoughts.

**Comment by steelbrain** [comment]
Actually, the linting **is** deferred until you stop typing, and if you think the timeout is a little too low, maybe ask the core folks about it. I agree that it would make more sense to have it centralized, therefore we have made it centralized, by not handling the deferring ourselves and leaving it to the core. :)
We are using this event from the core to detect if user stops typing
```js
atom.workspace.getActiveTextEditor().onDidStopChanging(function(e){
  console.log("User Stopped Changing")
})
```

**Comment by steelbrain** [comment]
Closing this for now, feel free to discuss more.

**Comment by steelbrain** [comment]
On a second thought, what do you guys think? @atom-community/linter-contributors

**Comment by adambuczynski** [comment]
I see, I misunderstood what you wrote earlier.

Do you think it should follow the same timeout throughout though? Maybe some things should be triggered right after you stopped typing, but others (like linting in my opinion) should wait for at least 1-2 seconds, because I might just be pausing briefly to consider the code I just wrote, before continuing on the same line. Seeing all the linting errors appear is very distracting at that point.

**Comment by gold** [comment]
I'm using Atom 1.17.0, and latest linter v2.14 and linter-eslint v8.1.7.

I cannot code with incessant nanny state popups telling me I'm missing a semi or I have trailing space because i hit the space bar and haven't typed another character yet.

The indicators in the tray at the bottom of the editor coupled with a red (error) yellowish (warning) glyph in the gutter are more than sufficient to indicate that there are linter warnings/errors.

Hovering over the line with the cursor will spawn the tooltip with all of the info i could use. This works perfectly because I can control exactly when or if I wish to see the detailed help.

Currently, the behavior is horrible: trying to help me for my own good -- whether i like it or not. The road to hell is paved with good intentions.

This is a Microsoft philosophy and it's one of the reasons, after I tried vscode, to switch back to Atom.

I am now forced to throw the baby out with the bath water because I cannot disable the continual tooltips popping up, obscuring part of the lines I wish to see -- while I'm typing!

No linter will ever bother me again while I type; I will issue the linter separately from the terminal periodically.

**Comment by adamreisnz** [comment]
@gold you are probably looking for this setting in the `linter-ui-default` package:

![image](https://cloud.githubusercontent.com/assets/490562/26266017/a0b79942-3d37-11e7-932c-ae7f2b8a01c2.png)

And I would also recommend turning off this one, in the `linter` base package:

![image](https://cloud.githubusercontent.com/assets/490562/26266065/dce85df2-3d37-11e7-9a16-fe74e77ac4f7.png)

**Comment by gold** [comment]
Unfortunately for me, I had looked at the two linter packages that I thought covered all the relevant options. linter description: "A Base Linter with Cow Powers" and I though that it was *the* base linter package dependency for linter-eslint (as well as all other individual, language-based linter packages).

I had overlooked the "linter-ui-default" package -- and impetuously revealed an emotional outburst in this forum lol.

Thanks to adamreisnz for setting me straight; I'm now back in a completely satisfied state with Atom!
